File: c++/NQueens.cpp
/**
 * @file NQueens.cpp
 * @brief Solution to the N-Queens problem using backtracking
 * @author Hacktoberfest Contributor
 * @date October 2025
 *
 * Problem: Place N chess queens on an N×N chessboard so that no two queens threaten each other.
 * A queen can attack horizontally, vertically, and diagonally.
 *
 * Algorithm: Backtracking
 * - Place queens one row at a time
 * - For each row, try placing queen in each column
 * - Check if placement is safe (no conflicts with previously placed queens)
 * - If safe, recursively place queens in next row
 * - If no valid placement found, backtrack and try different position
 *
 * Time Complexity: O(N!) - trying N positions in first row, N-1 in second, etc.
 * Space Complexity: O(N²) for the board + O(N) for recursion stack
 */

#include <iostream>
#include <vector>
#include <string>
using namespace std;

class NQueens {
private:
    /**
     * @brief Check if placing queen at board[row][col] is safe
     * @param board Current board configuration
     * @param row Row to place queen
     * @param col Column to place queen
     * @param n Board size
     * @return true if position is safe, false otherwise
     *
     * We only need to check:
     * 1. Same column (upper rows)
     * 2. Upper left diagonal
     * 3. Upper right diagonal
     * (No need to check rows below as we fill top to bottom)
     */
    bool isSafe(vector<string>& board, int row, int col, int n) {
        // Check column above
        for (int i = 0; i < row; i++) {
            if (board[i][col] == 'Q') {
                return false;
            }
        }

        // Check upper left diagonal
        for (int i = row - 1, j = col - 1; i >= 0 && j >= 0; i--, j--) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        // Check upper right diagonal
        for (int i = row - 1, j = col + 1; i >= 0 && j < n; i--, j++) {
            if (board[i][j] == 'Q') {
                return false;
            }
        }

        return true;
    }

    /**
     * @brief Optimized safety check using arrays to track columns and diagonals
     * @param cols Array tracking which columns have queens
     * @param diag1 Array tracking which left diagonals have queens
     * @param diag2 Array tracking which right diagonals have queens
     * @param row Current row
     * @param col Current column
     * @param n Board size
     * @return true if position is safe, false otherwise
     */
    bool isSafeOptimized(vector<bool>& cols, vector<bool>& diag1,
                         vector<bool>& diag2, int row, int col, int n) {
        // For left diagonal: row - col is constant (offset by n for positive index)
        // For right diagonal: row + col is constant
        return !cols[col] && !diag1[row - col + n] && !diag2[row + col];
    }

    /**
     * @brief Recursive backtracking function to solve N-Queens
     * @param board Current board state
     * @param row Current row to place queen
     * @param n Board size
     * @param solutions Vector to store all solutions
     */
    void solve(vector<string>& board, int row, int n, vector<vector<string>>& solutions) {
        // Base case: all queens placed successfully
        if (row == n) {
            solutions.push_back(board);
            return;
        }

        // Try placing queen in each column of current row
        for (int col = 0; col < n; col++) {
            if (isSafe(board, row, col, n)) {
                // Place queen
                board[row][col] = 'Q';

                // Recursively place queens in next rows
                solve(board, row + 1, n, solutions);

                // Backtrack: remove queen and try next position
                board[row][col] = '.';
            }
        }
    }

    /**
     * @brief Optimized recursive backtracking using tracking arrays
     * @param board Current board state
     * @param row Current row
     * @param n Board size
     * @param cols Column tracking array
     * @param diag1 Left diagonal tracking array
     * @param diag2 Right diagonal tracking array
     * @param solutions Vector to store all solutions
     */
    void solveOptimized(vector<string>& board, int row, int n,
                       vector<bool>& cols, vector<bool>& diag1, vector<bool>& diag2,
                       vector<vector<string>>& solutions) {
        if (row == n) {
            solutions.push_back(board);
            return;
        }

        for (int col = 0; col < n; col++) {
            if (isSafeOptimized(cols, diag1, diag2, row, col, n)) {
                // Place queen
                board[row][col] = 'Q';
                cols[col] = true;
                diag1[row - col + n] = true;
                diag2[row + col] = true;

                // Recurse
                solveOptimized(board, row + 1, n, cols, diag1, diag2, solutions);

                // Backtrack
                board[row][col] = '.';
                cols[col] = false;
                diag1[row - col + n] = false;
                diag2[row + col] = false;
            }
        }
    }

public:
    /**
     * @brief Solve N-Queens problem and return all solutions
     * @param n Size of the board (n x n)
     * @return Vector of all valid board configurations
     */
    vector<vector<string>> solveNQueens(int n) {
        vector<vector<string>> solutions;
        vector<string> board(n, string(n, '.'));

        solve(board, 0, n, solutions);
        return solutions;
    }

    /**
     * @brief Optimized solution using tracking arrays for O(1) safety checks
     * @param n Size of the board
     * @return Vector of all valid board configurations
     */
    vector<vector<string>> solveNQueensOptimized(int n) {
        vector<vector<string>> solutions;
        vector<string> board(n, string(n, '.'));

        // Tracking arrays for O(1) conflict detection
        vector<bool> cols(n, false);           // Track occupied columns
        vector<bool> diag1(2 * n, false);      // Track occupied left diagonals
        vector<bool> diag2(2 * n, false);      // Track occupied right diagonals

        solveOptimized(board, 0, n, cols, diag1, diag2, solutions);
        return solutions;
    }

    /**
     * @brief Print a single board solution
     * @param board Board configuration to print
     */
    void printBoard(const vector<string>& board) {
        for (const string& row : board) {
            cout << row << endl;
        }
        cout << endl;
    }

    /**
     * @brief Print all solutions
     * @param solutions Vector of all board configurations
     */
    void printAllSolutions(const vector<vector<string>>& solutions) {
        cout << "Total solutions: " << solutions.size() << endl << endl;

        for (int i = 0; i < solutions.size(); i++) {
            cout << "Solution " << (i + 1) << ":" << endl;
            printBoard(solutions[i]);
        }
    }
};

/**
 * @brief Example usage and test cases
 */
int main() {
    NQueens solver;

    // Test case 1: 4-Queens (classic small example)
    cout << "=== 4-Queens Problem ===" << endl;
    vector<vector<string>> solutions4 = solver.solveNQueens(4);
    solver.printAllSolutions(solutions4);

    // Test case 2: 8-Queens (standard chess board)
    cout << "\n=== 8-Queens Problem ===" << endl;
    vector<vector<string>> solutions8 = solver.solveNQueensOptimized(8);
    cout << "Total solutions for 8-Queens: " << solutions8.size() << endl;

    // Print first solution only (92 solutions total is too many)
    if (!solutions8.empty()) {
        cout << "\nFirst solution:" << endl;
        solver.printBoard(solutions8[0]);
    }

    // Test case 3: 1-Queen (trivial case)
    cout << "=== 1-Queen Problem ===" << endl;
    vector<vector<string>> solutions1 = solver.solveNQueens(1);
    solver.printAllSolutions(solutions1);

    // Test case 4: 3-Queens (impossible case)
    cout << "=== 3-Queens Problem ===" << endl;
    vector<vector<string>> solutions3 = solver.solveNQueens(3);
    if (solutions3.empty()) {
        cout << "No solution exists for 3-Queens" << endl;
    } else {
        solver.printAllSolutions(solutions3);
    }

    return 0;
}

File: javascript/nQueens.js
/**
 * N-Queens Problem
 * 
 * Description: Place N chess queens on an N×N chessboard so that no two queens
 * threaten each other. Uses backtracking to find all possible solutions.
 * 
 * Time Complexity: O(N!)
 * Space Complexity: O(N²)
 * 
 * Use Cases:
 * - Constraint satisfaction problems
 * - Resource allocation with conflicts
 * - Pattern generation and puzzle solving
 * - Algorithm education and demonstration
 * 
 * Example:
 * Input: n = 4
 * Output: [[".Q..","...Q","Q...","..Q."], ["..Q.","Q...","...Q",".Q.."]]
 */

function solveNQueens(n) {
    // Input validation
    if (!Number.isInteger(n) || n < 1) {
        throw new Error('Invalid input: n must be a positive integer');
    }

    const solutions = [];
    const board = Array.from({ length: n }, () => Array(n).fill('.'));
    const cols = new Set();
    const diag1 = new Set(); // row - col
    const diag2 = new Set(); // row + col

    function backtrack(row) {
        if (row === n) {
            solutions.push(board.map(r => r.join('')));
            return;
        }

        for (let col = 0; col < n; col++) {
            const d1 = row - col;
            const d2 = row + col;

            if (cols.has(col) || diag1.has(d1) || diag2.has(d2)) {
                continue;
            }

            // Place queen
            board[row][col] = 'Q';
            cols.add(col);
            diag1.add(d1);
            diag2.add(d2);

            backtrack(row + 1);

            // Remove queen (backtrack)
            board[row][col] = '.';
            cols.delete(col);
            diag1.delete(d1);
            diag2.delete(d2);
        }
    }

    backtrack(0);
    return solutions;
}

// Count total number of solutions
function totalNQueens(n) {
    let count = 0;
    const cols = new Set();
    const diag1 = new Set();
    const diag2 = new Set();

    function backtrack(row) {
        if (row === n) {
            count++;
            return;
        }

        for (let col = 0; col < n; col++) {
            const d1 = row - col;
            const d2 = row + col;

            if (cols.has(col) || diag1.has(d1) || diag2.has(d2)) {
                continue;
            }

            cols.add(col);
            diag1.add(d1);
            diag2.add(d2);

            backtrack(row + 1);

            cols.delete(col);
            diag1.delete(d1);
            diag2.delete(d2);
        }
    }

    backtrack(0);
    return count;
}

// Find first solution only (faster)
function findFirstSolution(n) {
    const board = Array.from({ length: n }, () => Array(n).fill('.'));
    const cols = new Set();
    const diag1 = new Set();
    const diag2 = new Set();

    function backtrack(row) {
        if (row === n) {
            return true;
        }

        for (let col = 0; col < n; col++) {
            const d1 = row - col;
            const d2 = row + col;

            if (cols.has(col) || diag1.has(d1) || diag2.has(d2)) {
                continue;
            }

            board[row][col] = 'Q';
            cols.add(col);
            diag1.add(d1);
            diag2.add(d2);

            if (backtrack(row + 1)) {
                return true;
            }

            board[row][col] = '.';
            cols.delete(col);
            diag1.delete(d1);
            diag2.delete(d2);
        }

        return false;
    }

    if (backtrack(0)) {
        return board.map(r => r.join(''));
    }
    return null;
}

// Visualize solution
function visualizeSolution(solution) {
    console.log('\n' + '='.repeat(solution[0].length * 4 + 1));
    for (let row of solution) {
        console.log('| ' + row.split('').join(' | ') + ' |');
        console.log('|' + '---+'.repeat(row.length).slice(0, -1) + '|');
    }
    console.log('='.repeat(solution[0].length * 4 + 1));
}

// Get queen positions as coordinates
function getQueenPositions(solution) {
    const positions = [];
    for (let row = 0; row < solution.length; row++) {
        for (let col = 0; col < solution[row].length; col++) {
            if (solution[row][col] === 'Q') {
                positions.push([row, col]);
            }
        }
    }
    return positions;
}

// Check if a configuration is valid
function isValidConfiguration(board) {
    const n = board.length;
    const queens = [];

    // Find all queens
    for (let i = 0; i < n; i++) {
        for (let j = 0; j < n; j++) {
            if (board[i][j] === 'Q') {
                queens.push([i, j]);
            }
        }
    }

    // Check all pairs
    for (let i = 0; i < queens.length; i++) {
        for (let j = i + 1; j < queens.length; j++) {
            const [r1, c1] = queens[i];
            const [r2, c2] = queens[j];

            // Same row, column, or diagonal
            if (r1 === r2 || c1 === c2 || 
                Math.abs(r1 - r2) === Math.abs(c1 - c2)) {
                return false;
            }
        }
    }

    return true;
}

// Optimized using bitmasking
function solveNQueensBitwise(n) {
    const solutions = [];
    const board = Array.from({ length: n }, () => Array(n).fill('.'));

    function backtrack(row, cols, diag1, diag2) {
        if (row === n) {
            solutions.push(board.map(r => r.join('')));
            return;
        }

        // Available positions (bits set to 1)
        let available = ((1 << n) - 1) & ~(cols | diag1 | diag2);

        while (available) {
            const pos = available & -available; // Rightmost set bit
            const col = Math.log2(pos);
            
            board[row][col] = 'Q';
            backtrack(
                row + 1,
                cols | pos,
                (diag1 | pos) << 1,
                (diag2 | pos) >> 1
            );
            board[row][col] = '.';

            available &= available - 1; // Clear rightmost set bit
        }
    }

    backtrack(0, 0, 0, 0);
    return solutions;
}

// Performance comparison
function comparePerformance(n) {
    console.log(`\nSolving ${n}-Queens problem:\n`);

    console.time('Standard backtracking');
    const solutions1 = solveNQueens(n);
    console.timeEnd('Standard backtracking');

    console.time('Bitwise optimization');
    const solutions2 = solveNQueensBitwise(n);
    console.timeEnd('Bitwise optimization');

    console.log(`Total solutions: ${solutions1.length}`);
    console.log(`Solutions match: ${solutions1.length === solutions2.length}`);
}

// Test cases
console.log('4-Queens solutions:');
const solutions = solveNQueens(4);
solutions.forEach((solution, i) => {
    console.log(`\nSolution ${i + 1}:`);
    visualizeSolution(solution);
});

console.log('\nTotal 8-Queens solutions:', totalNQueens(8)); // 92

console.log('\nFirst solution for 8-Queens:');
const first = findFirstSolution(8);
if (first) visualizeSolution(first);

comparePerformance(10);

module.exports = { 
    solveNQueens, 
    totalNQueens, 
    findFirstSolution,
    visualizeSolution,
    getQueenPositions,
    isValidConfiguration,
    solveNQueensBitwise
};
